### PR TITLE
SEP-38: distinguish sell/buy delivery methods for fee calculation

### PR DESCRIPTION
**Pull request by marcelosalloum**
### What

1. Replace `delivery_methods` with `buy_delivery_methods` and `sell_delivery_methods`.
2. Update request parameters of `GET /price`, `GET /prices` and `POST /quote` endpoints so the fee can be accurately calculated with the new parameters provided:
    * `sell_delivery_method`
    * `buy_delivery_method`
    * `country_code`
2. Add `country_codes` parameter to the `GET /info` endpoint
4. Fix request example for `GET /quote`.

### CODE REVIEW

File: ecosystem/sep-0038.md
Comment (leighmcculloch) [inline ecosystem/sep-0038.md:187]: I'm hesitant for us to include a SEP name in here. Can we include an `intent` or similar instead? For example, `"intent": "transfer"` for the transfer API SEPs SEP-6/24, `"intent": "cross_border_payment"` for SEP-31. There are probably better names.

**Review by leighmcculloch — COMMENTED** [review]
A question inline.

File: ecosystem/sep-0038.md
Comment (leighmcculloch) [reply ecosystem/sep-0038.md:187]: @JakeUrban wondering what your thoughts are on this.

**Review by JakeUrban — COMMENTED** [review]
I think we may need to add an additional field used to communicate how the client would like funds to be delivered to the user. We can't redefine `delivery_method` to be used for both because there are two deliveries of funds in a SEP-31 transaction, sender --> anchor and anchor --> receiver.

In general, we need to ensure the anchor has all the information needed to calculate their necessary margin. I think having the service provided (`sep`), rails used for both flows (`delivery_method`), and country of operation (`country_code`) is a good start. If an anchor requires other information, we can try and integrate that information in the future.

File: ecosystem/sep-0038.md
Comment (marcelosalloum) [reply ecosystem/sep-0038.md:187]: Regarding `"intent"` vs `"sep"`, I think both have downsides: in the future, we could end up having different protocols for "cross_border_payment" or different kinds of quote-related actions inside the same SEP, so both `&sep=6` or `&intent=cross_border_payment` could become insufficient to describe the intended action. TBH I like the "intent" key for it being more generic and easier to expand, but I also like the values "sep6" and "sep31" as they give a precise definition of what it'll be used for and they're also easier to expand with edge cases.

---

> `delivery_method` is only used for off-chain --> on-chain (i.e. deposit) flows, so we may need another attribute for the opposite flow, describing how funds are delivered to the user.

@JakeUrban: maybe we could tweak the description of delivery_methods a bit so it can be used for both buy_asset and sell_asset:
* Update the description of delivery_methods to say `An array of objects describing the methods a client can use to deliver funds to or withdraw funds from the anchor.`
* Update the `/quote`, `/price` and `/prices` request parameters to have `buy_delivery_method` and `sell_delivery_method` as parameters that should be used when buy_asset or sell_asset are non-Stellar assets.

Comment (JakeUrban) [reply ecosystem/sep-0038.md:187]: > Update the description of delivery_methods to say An array of objects describing the methods a client can use to deliver funds to or withdraw funds from the anchor.

I think this approach would make the spec more confusing and make it easier to pass semantically invalid `delivery_method` values. For example, a client could potentially pass `cash_pickup` for an off-chain --> on-chain transaction, when the appropriate value would've been `cash_dropoff`.

From a design perspective, the methods a client uses to deliver funds to the anchor are disjoint from the methods an anchor supports for delivering funds to the client/user. Why describe both sets of methods under the same object?

What if we changed the existing `delivery_methods` object & parameter to be `client_delivery_method` and added a `anchor_delivery_method` object & parameter? So `/quote`, `/price`, and `/prices` requests would include a `client_delivery_method` if it was an off-chain --> on-chain exchange and an `anchor_delivery_method` if it was an on-chain --> off-chain exchange.

Comment (marcelosalloum) [reply ecosystem/sep-0038.md:202]: @JakeUrban: I applied your suggestion in https://github.com/stellar/stellar-protocol/pull/989/commits/533d52fa85965b55d87ec71b3388c0e0292278aa. 

Q: should this trigger a major version increment from `1.2.0` to `2.0.0` or are we ok going to 1.3.0?

Comment (JakeUrban) [reply ecosystem/sep-0038.md:202]: > Q: should this trigger a major version increment from 1.2.0 to 2.0.0 or are we ok going to 1.3.0?

Thats a good question. Technically this is a breaking change (renaming an existing field) but since the SEP is in a draft state I don't think we need to strictly adhere to semantic versioning. @leighmcculloch curious on your thoughts here.

Comment (leighmcculloch) [reply ecosystem/sep-0038.md:202]: I'd try and keep the SEP version under 1 until it is in an Active state. We have no rules or process for this, but I've found that useful since v0 is a break anything you want version in regards to semantic versioning.

**Review by JakeUrban — APPROVED** [review]
Pending the version change, this LGTM, thanks @marcelosalloum

File: ecosystem/sep-0038.md
Comment (leighmcculloch) [inline ecosystem/sep-0038.md:155]: If these are always for deposit, could we use the term `deposit_delivery_methods`, and similar for anchors, `withdrawal_delivery_methods`? That builds on language we're already using for these two operations, where as `client` and `anchor` requires the reader to infer that during deposit they are the deliverer and therefore need to use this field.

cc @JakeUrban As I know you originally suggested client/anchor here.

Comment (JakeUrban) [reply ecosystem/sep-0038.md:155]: `deposit` and `withdraw` work too, but SEP-31 doesn't use that language. Maybe its still better than `client/anchor` since in SEP-31 the client is referred to as an anchor. I'm good with either 👍

Comment (leighmcculloch) [reply ecosystem/sep-0038.md:155]: `send` / `receive`?

Comment (marcelosalloum) [reply ecosystem/sep-0038.md:155]: What about `buy`/`sell`? They will be used along with parameters like buy_asset, buy_amount, sell_asset and sell_amount

File: ecosystem/sep-0038.md
Comment (leighmcculloch) [inline ecosystem/sep-0038.md:131]: It appears there is no `delivery_methods` anymore.
```suggestion
Object Schema for `sell_delivery_methods` and `buy_delivery_methods`.
```

**Review by leighmcculloch — APPROVED** [review]
I think this looks good. The buy/sell language works.

The PR title is not particularly specific. I suggest changing the title of the PR, and therefore the title of the final squashed commit, to:
> SEP-38: distinguish sell/buy delivery methods for fee calculation

File: ecosystem/sep-0038.md
Comment (marcelosalloum) [reply ecosystem/sep-0038.md:131]: Good suggestion!

File: ecosystem/sep-0038.md
Comment (JakeUrban) [inline ecosystem/sep-0038.md:206]: Can we avoid the terms `deposit`? That way we limit the number of terms we use when describing the movement of funds.

```suggestion
`sell_delivery_method` | string | (optional) One of the `name` values specified by the `sell_delivery_methods` array for the associated asset returned from [`GET /info`](#info). Must be provided if the user is delivering an off-chain asset to the anchor.
```

File: ecosystem/sep-0038.md
Comment (JakeUrban) [inline ecosystem/sep-0038.md:207]: Same thing here, and for the attributes on the `/price` endpoint.

```suggestion
`buy_delivery_method` | string | (optional) One of the `name` values specified by the `buy_delivery_methods` array for the associated asset returned from [`GET /info`](#info). Must be provided if the user intends to receive an off-chain asset from the anchor.
```

**Review by JakeUrban — CHANGES_REQUESTED** [review]
Very minor change, but otherwise LGTM! Once the syntax is changed I'll approve

File: ecosystem/sep-0038.md
Comment (marcelosalloum) [reply ecosystem/sep-0038.md]: Fixed in https://github.com/stellar/stellar-protocol/pull/989/commits/4256e69a463f879755c70b35e3032e28a3643391

File: ecosystem/sep-0038.md
Comment (marcelosalloum) [reply ecosystem/sep-0038.md]: Fixed in https://github.com/stellar/stellar-protocol/pull/989/commits/4256e69a463f879755c70b35e3032e28a3643391

**Review by JakeUrban — APPROVED** [review]

**Comment by marcelosalloum** [comment]
Hey @JakeUrban, the review was dismissed when I bumped the SEP update date in https://github.com/stellar/stellar-protocol/pull/989/commits/c304ac799ba2b15ac1c0ae109a6326ba0303b88e. Could you approve again?

**Review by JakeUrban — APPROVED** [review]

**Pull request merged**
